c-dependencies/js-compute-runtime/Makefile

```diff
@@ -3,36 +3,41 @@ WIZER ?= wizer
 
 CXX_OPT ?= -O2
 
-MODE=release
-CARGO_FLAG=--release
 ifdef DEBUG
-  MODE=debug
-  CARGO_FLAG=
+  MODE := debug
+  CARGO_FLAG :=
   CXX_OPT := $(CXX_OPT) -DDEBUG -DJS_DEBUG
+  Q :=
+else
+  MODE=release
+  CARGO_FLAG=--release
+  Q := @
 endif
 
-ROOT_SRC?=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))/..
-SM_SRC=$(ROOT_SRC)/spidermonkey/$(MODE)/
-SM_OBJ=$(SM_SRC)lib/*.o
-SM_OBJ+=$(SM_SRC)lib/*.a
-FSM_SRC=$(ROOT_SRC)/js-compute-runtime/
+ROOT_SRC ?= $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))/..
+SM_SRC := $(ROOT_SRC)/spidermonkey/$(MODE)/
+SM_OBJ := $(SM_SRC)lib/*.o
+SM_OBJ += $(SM_SRC)lib/*.a
+FSM_SRC := $(ROOT_SRC)/js-compute-runtime/
 
 WASI_CXX ?= /opt/wasi-sdk/bin/clang++
 
 
 CXX_FLAGS := -std=gnu++17 -Wall -Werror -Qunused-arguments -fno-sized-deallocation -fno-aligned-new -mthread-model single -fPIC -fno-rtti -fno-exceptions -fno-math-errno -pipe -fno-omit-frame-pointer -funwind-tables
+DEFINES ?=
 LD_FLAGS := -Wl,-z,noexecstack -Wl,-z,text -Wl,-z,relro -Wl,-z,nocopyreloc -Wl,-z,stack-size=1048576 -Wl,--stack-first
 
-.PHONY: all
+.PHONY: all clean compile_commands.json
 
 all: js-compute-runtime.wasm
 
 compiler_flags:
-	echo '$(CXX_OPT)' | cmp -s - $@ || echo '$(CXX_OPT)' > $@
+	echo '$(CXX_OPT) $(CXX_FLAGS)' | cmp -s - $@ || echo '$(CXX_OPT) $(CXX_FLAGS)' > $@
 
+WASM_STRIP = true
 ifeq (,$(findstring g,$(CXX_OPT)))
 ifneq (,$(shell which wasm-opt))
-WASM_STRIP = wasm-opt --strip-debug -o $@ $@
+WASM_STRIP = wasm-opt --strip-debug -o $1 $1
```

Comment:
Does `$1` work here? I see you've changed the usage of this variable to a `$(call)`, but given that my reading of the GNU make manual ([call function](https://www.gnu.org/software/make/manual/html_node/Call-Function.html), [variable references](https://www.gnu.org/software/make/manual/html_node/Reference.html)) is that you'd need to use `$(1)` rather than `$1` here.

I'm not sure either change is necessary anyway: `$@` means the target of the current rule, which should be the right thing here, shouldn't it?

Reply:
For variables with only a single character name, it's fine to use them without parens. This makefile behaves the same for calls to both `ECHO` and `ECHO2`:

```make
ECHO = echo $1                                                                       
ECHO2 = echo $(1)                                                                    
                                                                                     
foo: ci/*.sh                                                                         
        $(call ECHO,$<)                                                              
        $(call ECHO2,$<) 
```

I was a little bit uneasy leaving it as `$@` because it makes assumptions about where the `WASM_STRIP` macro will be  used (always in a rule context), while being explicit about it being a macro and using `$1` felt like it encoded the intent a little better.

I sort of wish that make didn't allow macros to be introduced with `=` :(

Reply:
Ah, I missed the last paragraph of the variable references section. Got it.

I wonder, could the conditional be shoved into the rule directly in a way that's more clear? I dunno, maybe like this:

```sh
	test -z "$(findstring g,$(CXX_OPT))" && which wasm-opt >/dev/null && wasm-opt --strip-debug -o $@ $@
```

I'm not convinced that looking for the single letter "g" makes sense here either, but maybe it's best to leave that alone. I suppose it works if you pass either `-Og` or `-g`?

Reply:
It'll definitely for unrelated occurrences of `g` in `CXX_OPT`. We could search for explicitly `-g`, but that will still match substrings.